doc/dev/dialects.rst

```diff
@@ -0,0 +1,422 @@
+
+Lastly, let's also add an operation to our dialect (again in the same file), which will allow us to
+run a small example at the end.
+
+.. code-block:: tablegen
+
+    include "mlir/IR/OpBase.td"
+
+    class OpenQASM_Op<string nameInIR> : Op<OpenQASMDialect, nameInIR, []>;
+
+    def RZGate : OpenQASM_Op<"RZ"> {
+        let summary = "A single-qubit rotation around the Z-axis by an angle θ.";
+
+        let arguments = (ins
+            F64:$theta,
+            QubitType:$qubit
+        );
+
+        let results = (outs
+        );
+
+        let assemblyFormat = [{
+            `(` $theta `)` $qubit attr-dict `:` type($qubit)
+        }];
+    }
+
+Operations are primarily defined via their *arguments* and *results*. In the IR, argument & result
+values are what organize operations into a graph (the so-called SSA graph), which encodes the flow
+of data through the program. The MLIR guide
+`Understanding the IR Structure <https://mlir.llvm.org/docs/Tutorials/UnderstandingTheIRStructure>`_
+can be helpful in obtaining a deeper understanding of this concept.
+
+Further, operations can define nested regions with additional operations. How the nested region
+will be executed is entirely up to the operation and its lowering mechanism. The concept of nesting
+operations is used in many places in MLIR. The built-in operations ``module`` and ``func`` are
+themselves just implemented as regular operations with a nested region. The structured control flow
+dialect (SCF) also uses it to represent branching and looping in a much more intuitive fashion
+than LLVM.
+
+Lastly, we also defined a custom syntax (``assemblyFormat``) for our operation. MLIR provides two ways
+of representing operations in its textual assembly format:
+
+- **generic assembly format**: This format is a one-to-one mapping from how MLIR objects are
+  represented in memory, and contains all necessary information to uniquely represent an MLIR
+  program with it. As a consequence, this format can be used to parse and print operations from any
+  dialect, even unknown ones!
+
+  The generic assembly format can be very useful for debugging, as it more truthfully represents the IR state. It
+also suffers less from crashing in the case of an invalid IR state.
+
+- **pretty assembly format**: This format can be fully customized (with some restrictions) by the
+  dialect designer. Generally speaking the IR can be much more human-readable when printed in this
+  form. Common improvements include imitating a particular syntax (e.g. indexed array access),
+  structuring operands into groups, and omitting redundant type information.
+
+More information on defining operations and other dialect objects can be found in the
+`dialects <https://mlir.llvm.org/docs/DefiningDialects>`_,
+`attributes & types <https://mlir.llvm.org/docs/DefiningDialects/AttributesAndTypes/>`_,
+and `operations <https://mlir.llvm.org/docs/DefiningDialects/Operations/>`_
+pages of the MLIR documentation.
+
+
```

Comment:
Maybe you should state which format did you use in the example above?

Reply:
The example above doesn't really use one or the other, since the assembly formats only come into play when representing IR in its textual form.

MLIR components always have a generic assembly form, and a pretty (custom) assembly form can be optionally defined if desired. Just above I say that we also defined a pretty format for our op when introducing the section about the different formats:
> Lastly, we also defined a custom syntax (``assemblyFormat``) for our operation.